lib/trie/hash.go

```diff
@@ -5,195 +5,285 @@ package trie
+// to the buffer given.
+func encodeBranch(b *branch, buffer *bytes.Buffer, parallel bool) (err error) {
 	if !b.dirty && b.encoding != nil {
-		return b.encoding, nil
+		_, err = buffer.Write(b.encoding)
+		return err
 	}
-	h.tmp.Reset()
 
 	encoding, err := b.header()
-	h.tmp.Write(encoding)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	h.tmp.Write(nibblesToKeyLE(b.key))
-	h.tmp.Write(common.Uint16ToBytes(b.childrenBitmap()))
+	buffer.Write(encoding)
+	buffer.Write(nibblesToKeyLE(b.key))
+	buffer.Write(common.Uint16ToBytes(b.childrenBitmap()))
 
 	if b.value != nil {
 		bytes, err := scale.Marshal(b.value)
 		if err != nil {
-			return nil, err
-		}
-		h.tmp.Write(bytes)
-	}
-
-	if h.parallel {
-		wg, _ := errgroup.WithContext(context.Background())
-		resBuff := make([][]byte, 16)
-		for i := 0; i < 16; i++ {
-			func(i int) {
-				wg.Go(func() error {
-					child := b.children[i]
-					if child == nil {
-						return nil
-					}
-
-					var err error
-					resBuff[i], err = encodeAndHash(child)
-					if err != nil {
-						return err
-					}
-					return nil
-				})
-			}(i)
-		}
-		if err := wg.Wait(); err != nil {
-			return nil, err
+			return err
 		}
+		buffer.Write(bytes)
+	}
+
+	if parallel {
+		return encodeChildsInParallel(b.children, buffer)
+	}
+	return encodeChildsSequentially(b.children, buffer)
+}
 
-		for _, v := range resBuff {
-			if v != nil {
-				h.tmp.Write(v)
+func encodeChildsInParallel(children [16]node, buffer *bytes.Buffer) (err error) {
+	type result struct {
+		index  int
+		buffer *bytes.Buffer
+		err    error
+	}
+
+	resultsCh := make(chan result)
+
+	for i, child := range children {
+		go func(index int, child node) {
+			buffer := encodingBufferPool.Get().(*bytes.Buffer)
+			buffer.Reset()
+			// buffer is put back in the pool after processing its
+			// data in the select block below.
+
+			err := encodeChild(child, buffer)
+
+			resultsCh <- result{
+				index:  index,
+				buffer: buffer,
+				err:    err,
 			}
+		}(i, child)
+	}
+
+	currentIndex := 0
+	resultBuffers := make([]*bytes.Buffer, len(children))
+	for range children {
+		result := <-resultsCh
+		if result.err != nil && err == nil { // only set the first error we get
+			err = result.err
 		}
-	} else {
-		for i := 0; i < 16; i++ {
-			if child := b.children[i]; child != nil {
-				scEncChild, err := encodeAndHash(child)
-				if err != nil {
-					return nil, err
-				}
-				h.tmp.Write(scEncChild)
+
+		resultBuffers[result.index] = result.buffer
+
+		// write as many completed buffers to the result buffer.
+		for currentIndex < len(children) &&
+			resultBuffers[currentIndex] != nil {
+			// note buffer.Write copies the byte slice given as argument
+			_, writeErr := buffer.Write(resultBuffers[currentIndex].Bytes())
+			if writeErr != nil && err == nil {
+				err = writeErr
 			}
+
+			encodingBufferPool.Put(resultBuffers[currentIndex])
+			resultBuffers[currentIndex] = nil
+
+			currentIndex++
 		}
```

Comment:
ok, maybe I'm just having trouble reading this code lol, but does this preserve the order that the children are written? eg. a child at index 0 should be written before one at index 1, etc

Reply:
Haha sorry I know the code is convoluted due to the use of pooled buffers and other buffers, for memory footprint.

`resultBuffers` is a slice of buffers (pointers to buffers), and we set the index in this `resultBuffers` slice using the index coming from the results channel. We then process from index 0 the buffers from `resultBuffers`, and as soon as a sequence of buffers is completed in order to release back buffers to the buffer pool asap.

Reply:
ok cool, I kinda understand now lol. as long as the tests pass it's good :p